interpreter/beam/beam.go

```diff
@@ -0,0 +1,510 @@
+	}
+
+	// "erts_atom_table" symbol is from:
+	// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/atom.c#L35
+	erts_atom_table, err := ef.LookupSymbolAddress("erts_atom_table")
+	if err != nil {
+		return nil, fmt.Errorf("symbol 'erts_atom_table' not found: %v", err)
+	}
+
+	interpRanges, err := info.GetSymbolAsRanges(libpf.SymbolName("process_main"))
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Do we need this if all the actual code is JITed?
+	if err = ebpf.UpdateInterpreterOffsets(support.ProgUnwindBEAM, info.FileID(), interpRanges); err != nil {
+		return nil, err
+	}
+
+	d := &beamData{
+		version:               otpVersion,
+		the_active_code_index: uint64(the_active_code_index),
+		r:                     uint64(r),
+		erts_atom_table:       uint64(erts_atom_table),
+	}
+
+	log.Infof("BEAM loaded, OTP version %d", otpVersion)
+
+	return d, nil
+}
+
+func (d *beamData) Attach(ebpf interpreter.EbpfHandler, pid libpf.PID, bias libpf.Address, rm remotememory.RemoteMemory) (interpreter.Instance, error) {
+	log.Infof("BEAM attaching, bias: 0x%x", bias)
+
+	data := C.BEAMProcInfo{
+		version:               C.u32(d.version),
+		bias:                  C.u64(bias),
+		r:                     C.u64(uint64(bias) + d.r),
+		the_active_code_index: C.u64(uint64(bias) + d.the_active_code_index),
+	}
+	if err := ebpf.UpdateProcData(libpf.BEAM, pid, unsafe.Pointer(&data)); err != nil {
+		return nil, err
+	}
+
+	return &beamInstance{
+		pid:       pid,
+		data:      d,
+		rm:        rm,
+		rangesPtr: bias + libpf.Address(d.r),
+		atomTable: bias + libpf.Address(d.erts_atom_table),
+	}, nil
+}
+
+func (d *beamData) Unload(_ interpreter.EbpfHandler) {
+}
+
+func (i *beamInstance) SynchronizeMappings(ebpf interpreter.EbpfHandler, _ reporter.SymbolReporter, pr process.Process, mappings []process.Mapping) error {
+	pid := pr.PID()
+
+	// Index into the active static `r` variable using each valid index (the size of the `ranges` struct is 32 bytes)
+	// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/beam_ranges.c#L62
+	// The max index is defined as 3 here: https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/code_ix.h#L70
+	for codeIndex := 0; codeIndex < 3; codeIndex++ {
+		activeRanges := i.rangesPtr + libpf.Address(32*codeIndex)
+
+		// Use offsets into the `ranges` struct to get the beginning of the array and the number of entries based on
+		// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/beam_ranges.c#L56-L61
+		modules := i.rm.Ptr(activeRanges)
+		n := i.rm.Uint64(activeRanges + libpf.Address(8))
+
+		low := i.rm.Ptr(modules)
+		high := i.rm.Ptr(modules + libpf.Address((n-1)*16+8))
+
+		log.Infof("Enabling BEAM for %#x - %#x", low, high)
+
+		// TODO: I think this is resulting in the following error (figure out why):
+		// ERRO[0004] Failed to handle new anonymous mapping for PID 375011: update: key already exists
+		prefixes, err := lpm.CalculatePrefixList(uint64(low), uint64(high))
```

Comment:
You need to track the previously added prefixes and update only add the changed ones, for an example, see: https://github.com/open-telemetry/opentelemetry-ebpf-profiler/blob/dd36674acb898f835a4232ea5e69150319b1c477/interpreter/nodev8/v8.go#L572